src/containers/org/components/PermissionsPanel.js

```diff
@@ -185,6 +196,41 @@ const PermissionsPanel = ({
     setLocalPermissions(permissions);
   };
 
+  const togglePermissions = () => {
+    if (isPermissionAssignedToAll) {
+      // remove permission from all properties
+      const updatedPermissions :Map<List<UUID>, Ace> = Map().withMutations((mutableMap) => {
+        properties.keySeq().forEach((id :UUID) => {
+          const key = List([dataSetId, id]);
+          const localAce :?Ace = localPermissions.get(key);
+          const updatedAcePermissions = Set(localAce?.permissions).delete(permissionType);
+          const updatedAce = (new AceBuilder())
+            .setPermissions(updatedAcePermissions)
+            .setPrincipal(principal)
+            .build();
+          mutableMap.set(key, updatedAce);
+        });
+      });
+      setLocalPermissions(updatedPermissions);
+    }
+    else {
+      // add permission to all properties
+      const updatedPermissions :Map<List<UUID>, Ace> = Map().withMutations((mutableMap) => {
+        properties.keySeq().forEach((id :UUID) => {
+          const key = List([dataSetId, id]);
+          const localAce :?Ace = localPermissions.get(key);
+          const updatedAcePermissions = Set(localAce?.permissions).add(permissionType);
+          const updatedAce = (new AceBuilder())
+            .setPermissions(updatedAcePermissions)
+            .setPrincipal(principal)
+            .build();
+          mutableMap.set(key, updatedAce);
+        });
+      });
+      setLocalPermissions(updatedPermissions);
```

Comment:
Are there any potential issues if one of these requests fails and a user tries again? What happens if a user ends up granting permissions for something that already exists or removing what doesn't exist

Reply:
`setLocalPermissions` is updating local state, i.e. `useState()`. Persisting the changes happens when they click the save button.

I have not yet added error handling for when permissions are being saved. In theory, if the request fails, nothing in the system has changed... redux store is the same, component state is the same... so I think clicking the save button again should be safe.

I'm not sure how the backend responds to the scenarios you mentioned. I am using the `SET` action type, not the `ADD` / `REMOVE` action type, so I believe the backend will just take what we pass it and overwrite what currently exists.